core-foundation/src/propertylist.rs

```diff
@@ -56,8 +58,142 @@ pub fn create_data(property_list: *const c_void, format: CFPropertyListFormat) -
     }
 }
 
+
+/// Trait for all subclasses of CFPropertyList.
+pub trait CFPropertyListSubClass<Raw>: TCFType<*const Raw> {
```

Comment:
We should be able to do `pub trait CFPropertyListSubClass: TCFType<*const Self> {` and avoid some redundancy in the implementations.

Reply:
Getting rid of the generic would be awesome. But I don't understand exactly how that would work. `Self` here is `CFString` or similar.

Reply:
Separate from this I'm working on trying to make the `ConcreteTypeRef` in `TCFType` a associated type instead of a generic. When/if that works out this can be greatly simplified.

Reply:
`impl CFPropertyListSubClass<::data::__CFData> for ::data::CFData` duplicates the type in two places. Using the snippet I provided, the resulting impl of `impl CFPropertyListSubClass for ::data::CFData` should work just fine: https://play.rust-lang.org/?gist=b89fc9fe403069fd843f6054897f21db&version=stable

Reply:
Yes, but `impl U<*const u32> for u32 {}` does not hold for us.

Our `CFString` does not impl `TCFType<*const CFString>`, it implements it for `TCFType<*const __CFString>`.

Reply:
The existance of both `CFStringRef` and `__CFString` makes it complicated in more than one situation for me. I can't cast between pointers in the downcast method because the compiler does not know that `CFStringRef` is a raw pointer. It is, and all `*Ref` are `*const _`, but the compiler does not know. Which forces me to reference the underlying `*const __CF<Foo>` types in order to show the compiler the casts are fine. Maybe it's possible to work around it in some way, but I have not found one yet.

Reply:
Oh, whoops. I didn't notice the underscores in the type name.

Reply:
Easy thing to do. Lots of types with very similar names here :)